builtin/credential/appgroup/user_id.go

```diff
@@ -0,0 +1,372 @@
+		lock.RUnlock()
+		return false, nil
+	} else if err := entry.DecodeJSON(&result); err != nil {
+		lock.RUnlock()
+		return false, err
+	}
+
+	// NumUses will be zero only if the usage limit was not set at all,
+	// in which case, the UserID will remain to be valid as long as it is not
+	// expired.
+	if result.NumUses == 0 {
+		lock.RUnlock()
+		return true, nil
+	}
+
+	// If the NumUses is non-zero, it means that its use-count should be updated
+	// in the storage. Switch the lock from a `read` to a `write` and update
+	// the storage entry.
+	lock.RUnlock()
+	b.userIDLocksMapLock.RUnlock()
+
+	b.userIDLocksMapLock.Lock()
+	defer b.userIDLocksMapLock.Unlock()
+
+	lock.Lock()
+	defer lock.Unlock()
+
+	// Lock switching may change the data. Refresh the contents.
+	result = userIDStorageEntry{}
+	if entry, err := s.Get(entryIndex); err != nil {
+		return false, err
+	} else if entry == nil {
+		return false, nil
+	} else if err := entry.DecodeJSON(&result); err != nil {
+		return false, err
+	}
+
+	// If there exists a single use left, delete the UserID entry from
+	// the storage but do not fail the validation request. Delete the
+	// UserIDs lock from the map of locks. Subsequest requests to use
+	// the same UserID will fail.
+	if result.NumUses == 1 {
+		if err := s.Delete(entryIndex); err != nil {
+			return false, err
+		}
+		// The storage entry for generic type is not created by any endpoints
+		// and it is not cleaned up in any other way. When the UserID belonging
+		// to the generic storage entry is getting invalidated, the entry should
+		// be deleted as well.
+		if selectorType == selectorTypeGeneric {
+			if err := b.deleteGenericEntry(s, selectorValue); err != nil {
+				return false, err
+			}
+		}
+		// Reset the lock that is used to modify the UserID's storage entry.
+		b.userIDLocksMap[userID] = nil
+	} else {
+		// If the use count is greater than one, decrement it and update the last updated time.
+		result.NumUses -= 1
+		result.LastUpdatedTime = time.Now().UTC()
+		if entry, err := logical.StorageEntryJSON(entryIndex, &result); err != nil {
+			return false, fmt.Errorf("failed to decrement the num_uses for user ID:%s", userID)
+		} else if err = s.Put(entry); err != nil {
+			return false, fmt.Errorf("failed to decrement the num_uses for user ID:%s", userID)
+		}
+	}
+
+	return true, nil
+}
+
+// registerUserIDEntry creates a new storage entry for the given UserID.
+// Successful creation of the storage entry results in the creation of a
+// lock in the map of locks maintained at the backend. The index into the
+// map is the UserID itself. During login, if the UserID supplied is not
+// having a corresponding lock in the map, the login attempt fails.
+func (b *backend) registerUserIDEntry(s logical.Storage, selectorType, selectorValue, userID string, userIDEntry *userIDStorageEntry) error {
+	b.userIDLocksMapLock.RLock()
+	if b.userIDLocksMap[userID] != nil {
+		return fmt.Errorf("user ID is already registered")
```

Comment:
I really strongly encourage you to look at `cubbyhole-the-world`'s token store and just allocate the locks ahead of time. It does mean that you may have to hit storage to load a current value but you're doing this anyways. This will make things noticeably simpler to reason about...for instance, no lock switch + refresh.